src/adaptors/tcp_adaptor.c

```diff
@@ -630,64 +780,92 @@ static void handle_connection_event(pn_event_t *e, qd_server_t *qd_server, void
         }
     }
     case PN_RAW_CONNECTION_CLOSED_READ: {
-        qd_log(log, QD_LOG_DEBUG, "[C%"PRIu64"] PN_RAW_CONNECTION_CLOSED_READ", conn->conn_id);
-        conn->q2_blocked = false;
-        handle_incoming_impl(conn, true);
+        qd_log(log, QD_LOG_DEBUG, "[C%"PRIu64"][L%"PRIu64"] PN_RAW_CONNECTION_CLOSED_READ",
+               conn->conn_id, conn->incoming_id);
         sys_mutex_lock(conn->activation_lock);
+        conn->q2_blocked = false;
         conn->raw_closed_read = true;
         sys_mutex_unlock(conn->activation_lock);
-        pn_raw_connection_close(conn->pn_raw_conn);
+        handle_incoming(conn, "PNRC_CLOSED_READ");
         break;
     }
     case PN_RAW_CONNECTION_CLOSED_WRITE: {
-        qd_log(log, QD_LOG_DEBUG, "[C%"PRIu64"] PN_RAW_CONNECTION_CLOSED_WRITE", conn->conn_id);
+        qd_log(log, QD_LOG_DEBUG,
+               "[C%"PRIu64"] PN_RAW_CONNECTION_CLOSED_WRITE",
+               conn->conn_id);
         sys_mutex_lock(conn->activation_lock);
         conn->raw_closed_write = true;
         sys_mutex_unlock(conn->activation_lock);
-        pn_raw_connection_close(conn->pn_raw_conn);
+        if (conn->ingress) {
```

Comment:
Why only on ingress?

I would expect that on CLOSED_WRITE we settle the incoming qdr delivery, i.e. outstream (since we can no longer forward any data received). That signals the peer that they should close their socket for read.

Reply:
I will try removing that conditiona.

Reply:
Removing the conditional causes data loss on the connector side.
On a connector/server connection the sequence goes:

- EOS seen. Call pnrc_write_close() for connector connection
- PNRC_CLOSED_WRITE event
- Call pnrc_connection_close()  <-- called because conditional was removed
- PNRC_CLOSED_READ event
- PNRC_READ and PNRC_WRITTEN to recover buffers
- PNRC_CONNECTION_DISCONNECTED

If connection_close is not called on the connector then the write_close() propagates to the server, the server closes the connection, the connector-side of the adaptor receives the server's last data, and then CLOSED_READ comes in and the connection is shut.

Reply:
Getting rid of the conditional AND the pn_connection_close() call seems to work correctly. Just remove the code block entirely.